internal/pwru/bpf_prog.go

```diff
@@ -0,0 +1,163 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2023 Leon Hwang.
+
+package pwru
+
+import (
+	"errors"
+	"fmt"
+	"log"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/btf"
+	"github.com/cilium/ebpf/link"
+	"golang.org/x/sys/unix"
+)
+
+func listBpfProgs(typ ebpf.ProgramType) ([]*ebpf.Program, error) {
+	var (
+		id  ebpf.ProgramID
+		err error
+	)
+
+	var progs []*ebpf.Program
+	for id, err = ebpf.ProgramGetNextID(id); err == nil; id, err = ebpf.ProgramGetNextID(id) {
+		prog, err := ebpf.NewProgramFromID(id)
+		if err != nil {
+			return nil, err
+		}
+
+		if prog.Type() == typ {
+			progs = append(progs, prog)
+		} else {
+			_ = prog.Close()
+		}
+	}
+
+	if err != nil && !errors.Is(err, unix.ENOENT) {
+		return nil, err
+	}
+
+	return progs, nil
+}
+
+func listFuncs(prog *ebpf.Program) ([]*btf.Func, error) {
+	info, err := prog.Info()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get program info: %w", err)
+	}
+
+	id, ok := info.BTFID()
+	if !ok {
+		return nil, fmt.Errorf("bpf program %s does not have BTF", info.Name)
+	}
+
+	handle, err := btf.NewHandleFromID(id)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get BTF handle: %w", err)
+	}
+	defer handle.Close()
+
+	spec, err := handle.Spec(nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get BTF spec: %w", err)
+	}
+
+	var funcs []*btf.Func
+	iter := spec.Iterate()
+	for iter.Next() {
```

Comment:
Do we actually need to iterate over all `Func`'s? The entry function is the first one, so maybe let's call `break` once the first `Func` is found.